#Перевірка статиcтичних гіпотез. Z-test.



0. Зчитайте дані з `data.csv` у змінну data, яка має тип pandas.DataFrame. Ми будемо далі працювати з цією змінною.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

In [2]:
data = pd.read_csv(r"C:\Users\User\Desktop\DATA LOVES ACADEMY\HOMETASKS\statistical_hypothesis\data.csv")

Запустіть код нижче. Для коректної робити всіх подальших методів дані мають бути у вигляді одновимірного `numpy.ndarray` вектора та мати розмірність (100,). Така розмірність означає, що дані - одновимірні (колонка), якщо б розмірність була (100, 1), то дані сприймаються методами як двовимірні (таблиця), хоча для нас виглядати можуть так само.

In [3]:
data = data.data.values
data.shape

(100,)

**Завдання 1**. (7 балів)
Зчитані дані - це сума покупок в доларах наших покупців на сайті протягом липня 2025 року.

До цього місяця, ми мали продажі в середньому на чек $20.

Необхідно зʼясувати, чи вийшло у нас статистично значущо **підвищити** середній чек за липень 2025?

Аби дати відповідь - ми проведемо z-test на рівні значущості $\alpha=0.05$ двома способами. В цьому завданні ми будемо виконувати обчислення "вручну" з використанням бібліотек numpy та scipy.stats подібно до прикладу в лекції. Для цього виконайте наступні 7 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть параметри задачі у змінні Python:
    - `sample_mean` - $\bar{x}$, середнє значення в вибірці
    - `population_mean` - $\mu_0$, середнє значення в популяції (тобто те, з яким ми порівнюємо середнє вибіркове значення)
    - `population_std` - $\sigma$, вибіркове стандартне відхилення, яке ми вважаємо, що є рівним ст. відх. популяції, адже маємо достатньо велику вибірку.
    - `sample_size` - $n$, розмір вибірки
    - `alpha` - $\alpha$ рівень значущості
    
    **Увага!** Для обчислення стандартного відхилення маємо скористатись функцією `np.std(your_dataframe, ddof=1)`. Чому так - розбираємо в лекції про t-test.


2. Визначте для цієї задачі:
    - якою є гіпотеза $H_0$
    - якою є альтернативна гіпотеза $H_a$
    - з яким типом тесту ми маємо справу - лівосторонній, правосторонній чи двосторонній.

3. Обчисліть стандартну помилку SE.
4. Розрахуйте z-статистику.
5. Знайдіть критичне z-значення з допомогою бібліотеки stats.
6. Обчисліть p-value з допомогою бібліотеки stats.
7. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$. Для прийняття рішення зробіть і порівняння z-статистики з критичним значенням, і проаналізуйте p-value.

In [17]:
# Задаємо всі відомі параметри
sample_mean = np.mean(data)                     # x̄ - середнє у вибірці
print(f"sample_mean: {sample_mean:.4f}")
population_mean = 20                            # μ₀ - продажі в середньому на чек
print(f"population_mean: {population_mean}")
population_std = np.std(data, ddof=1)           # σ - стандартне відхилення в популяції
print(f"population_std: {population_std:.4f}")
sample_size = data.size                         # n - розмір вибірки
print(f"sample_size: {sample_size}")
alpha = 0.05                                    # α - рівень значущості (ймовірність помилки I роду)
print(f"alpha: {alpha}")

sample_mean: 19.3776
population_mean: 20
population_std: 6.2536
sample_size: 100
alpha: 0.05


**Гіпотези**   

H₀: μ = 20  (середні продажі на чек)  
H₁: μ > 20 (середній чек виріс у липні 2025)  
Тип тесту: правосторонній (one-tailed)  

**Стандартна помилка (SE)**

In [9]:
# Формула: SE = σ / √n
standard_error = population_std / np.sqrt(sample_size)
print(f"SE = {standard_error:.3f}")

SE = 0.625


**Розраховуємо z-статистику**

In [10]:
# Формула: z = (x̄ - μ₀) / SE
z_statistic = (sample_mean - population_mean) / standard_error
print(f"z = {z_statistic:.3f}")

z = -0.995


**Знаходимо критичне значення**

In [19]:
# Для правостороннього тесту з α = 0.05
z_critical = stats.norm.ppf(1 - alpha)
print(f"Критичне z-значення = {z_critical:.3f}")

Критичне z-значення = 1.645


**Обчислюємо p-value**

In [21]:
# Для правостороннього тесту: p-value = P(Z > z_statistic)
p_value = 1 - stats.norm.cdf(z_statistic)
print(f"p-value = {p_value:.3f}")

p-value = 0.840


**Порівняння з критичним значенням**
Для **правостороннього** тесту область відхилення:
$$
Z \geq 1.645
$$
А в нас:
$$
z = -0.995
$$
Перевірка:
$$
-0.995 < 1.645
$$
 **z-статистика не потрапляє в область відхилення**, отже **$H_0$ не відхиляємо**.

**Аналіз p-value**

Правило прийняття рішення:
- якщо $p\text{-value} \leq \alpha$ → **відхиляємо** $H_0$
- якщо $p\text{-value} > \alpha$ → **не відхиляємо** $H_0$
У нас:
$$
p\text{-value} = 0.840 > 0.05 = \alpha
$$
 **$H_0$ не відхиляємо**.

**Завдання 2.** (3 бали)
Виконайте обчислення z-test з використанням бібліотеки statsmodels.

Отримайте z-статистику та р-значення.

Виведіть p-значення та зробіть висновок, чи ми маємо достатньо доказів, аби стверджувати, що середній чек зріс.

Чи зійшлись значення z-статистику та р-значення в цьому завданні з попередніми обчисленнями?

In [24]:
from statsmodels.stats import weightstats as stests

In [28]:
z_stat_sm, p_val_sm = stests.ztest(data,           # наші дані
                                    value=population_mean,  # μ₀ = 20
                                    alternative='larger')   # H₁: μ > 20

print(f"Результати z-тесту:")
print(f"  Z-статистика: {z_stat_sm:.3f}")
print(f"  P-value: {p_val_sm:.3f}")

Результати z-тесту:
  Z-статистика: -0.995
  P-value: 0.840


**Аналіз p-value**

Правило прийняття рішення:
- якщо $p\text{-value} \leq \alpha$ → **відхиляємо** $H_0$
- якщо $p\text{-value} > \alpha$ → **не відхиляємо** $H_0$
У нас:
$$
p\text{-value} = 0.840 > 0.05 = \alpha
$$
 $H_0$ не відхиляємо. Не достатньо доказів, аби стверджувати, що середній чек зріс

**Значення z-статистики та р-значення в цьому завданні з попередніми обчисленнями повністю зійшлись**